zachyazzie/CSE341-team
# Eb 12 6 addtest

Tests are now passing. I was able to get tests for the routes working

Zach, can you approve the two PRs in Github?

Here's what they do:
(1) First PR was to remove auth from routes and export app in index for test files to work
(2) Second PR includes modifications to test files to get them to run. Below is a summary of this PR:

**Changes made in Second PR**

**I followed Brother Dransfield's video, which:**

- Creates a helper function in a file called "routesFor.js" - this is for the tests that test for the existence of routes (is there a GET, POST, etc.) - if you watch the video it will make sense. 
- Adds a test to test for the existence of http routes in each of the four routes files - again, probably helps to watch the video.

**In addition to testing for the existence of routes, I also modified the route test files as such:**

- Modifies the GET tests in the test files that Ulises made to allow for async - I was getting timeouts otherwise
- In each of those test files, it also adds the Authorization header with the Access token (see note below on that)
- I commented out the other tests for now, since only one is required, and I was not sure if the test for routes would count, since it is not making a call to the database. You can see the syntax difference. There are a few ways to do this, but the way it was in there before, it was timing out. There are ways to increase the timer, apparently ignore the timer, but you can also just make the calls async, which is what I did, and it works.
- I also found a solution for the issue where multiple ports were being opened. You can find the solution I used [here](https://stackoverflow.com/questions/54422849/jest-testing-multiple-test-file-port-3000-already-in-use). Basically, in the root index.js file, it tests for the NODE_ENV variable. If the test is run, it will re-run the code here for each test file, thus attempting to open a port over and over. This just simply says, if the NODE_ENV variable = "test", skip the connect logic, since we are connecting before we begin the test anyway. 

I also installed supertest, which added the dependencies to the package.json file. You may already have those there. [Here's more on supertest](https://www.npmjs.com/package/supertest).

I also fixed a bug in the users controller I coded, so that was good - the test found it. Kind of cool.

Finally, because the authorization token is now successfully being passed into the test functions that access the database, I uncommented out again the "loaduser" lines of code.

With all of those changes, I authenticated, got my access token, added the access token to my .env file, and ran the tests and they all succeeded.

**Process to test**

1. npm start
2. go to localhost:8080/authorization/login --> should get an access token back (if haven't done this yet, you can log in with google)
3. Copy access token and paste in ACCESS_TOKEN in .env file
4. stop app because we made a change to a file
5. npm start
6. open another terminal, and npm test

**Issue with Render**
Because we are using the ACCESS_TOKEN env variable, we can only have one access token in render at a time. This effectively means that only Zach (since you have access to change the env variables in render) can test the tests (or we do it as a group) with render. Testing the routes with the api-docs interface won't be affected, because you can click "authorize" and enter the access token there. I'm not sure how else to handle that. Not sure if the grader requires to run the tests or if we are able to just demonstrate it in the video. I might ask that in the General tab.
